src/corenetwork/lteip/LteIp.cc

```diff
@@ -132,195 +135,226 @@ void LteIp::endService(cPacket *msg)
-            EV << "LteIp (UE): Wrong gate " << msg->getArrivalGate()->getName() << endl;
+            EV << "LteIp (UE): Wrong gate " << msg->getArrivalGate()->getName()
+                      << endl;
         }
     }
 
     if (getEnvir()->isGUI())
-    updateDisplayString();
+        updateDisplayString();
 }
 
-void LteIp::fromTransport(cPacket * transportPacket, cGate *outputgate)
-{
-    // Remove control info from transport packet
-    IPv4ControlInfo *ipControlInfo = check_and_cast<IPv4ControlInfo*>(transportPacket->removeControlInfo());
+void LteIp::fromTransport(cPacket * transportPacket, cGate *outputgate) {
+
+    auto packetTmp = check_and_cast<inet::Packet*>(transportPacket);
+    auto ipControlInfo = make_shared<Ipv4Header>(
+            *(packetTmp->popAtFront<Ipv4Header>()));
 
     //** Create IP datagram and fill its fields **
 
-    IPv4Datagram *datagram = new IPv4Datagram(transportPacket->getName());
+
+//    TODO: needs proper testing
+    inet::Packet *datagram = new inet::Packet(transportPacket->getName(), 0);
     datagram->setByteLength(IP_HEADER_BYTES);
     datagram->encapsulate(transportPacket);
 
     // set destination address
-    IPv4Address dest = ipControlInfo->getDestAddr();
-    datagram->setDestAddress(dest);
+    Ipv4Address dest = ipControlInfo->getDestAddress();
+    auto header = makeShared<Ipv4Header>(*datagram->peekAtFront<Ipv4Header>());
+    header->setDestAddress(dest);
+
+//    datagram->setDestAddress(dest);
 
     // set source address
-    IPv4Address src = ipControlInfo->getSrcAddr();
+    Ipv4Address src = ipControlInfo->getSrcAddress();
 
     // when source address was given, use it; otherwise use local Addr
-    if (src.isUnspecified())
-    {
+    if (src.isUnspecified()) {
         // find interface entry and use its address
-        IInterfaceTable *interfaceTable = getModuleFromPar<IInterfaceTable>(par("interfaceTableModule"), this);
+        IInterfaceTable *interfaceTable = getModuleFromPar<IInterfaceTable>(
+                par("interfaceTableModule"), this);
         // TODO: how do we find the LTE interface?
-        src = interfaceTable->getInterfaceByName("wlan")->ipv4Data()->getIPAddress();
+        src = interfaceTable->getInterfaceByName("wlan")->getIpv4Address();
         EV << "Local address used: " << src << endl;
+    } else {
+        EV << "Source address in control info from transport layer " << src
+                  << endl;
     }
-    else
-    {
-        EV << "Source address in control info from transport layer " << src << endl;
-    }
-    datagram->setSrcAddress(src);
+    header->setSrcAddress(src);
 
     // set other fields
-    datagram->setTypeOfService(ipControlInfo->getTypeOfService());
-    datagram->setTransportProtocol(ipControlInfo->getProtocol());
+    header->setTypeOfService(ipControlInfo->getTypeOfService());
+    // TODO: header immutable? may be dropped
+    datagram->popAtFront<Ipv4Header>();
+    datagram->insertAtFront(header);
```

Comment:
I think you shoud call trimAtFront() before being able to call insertAtFront() after calling popAtFront().

Reply:
Yes, call to popFront() added.